packages/instantsearch-ui-components/src/components/autocomplete/createAutocompleteStorage.ts

```diff
@@ -0,0 +1,139 @@
+import type { UsePropGetters } from './createAutocompletePropGetters';
+
+type CreateAutocompleteStorageParams = {
+  useEffect: (effect: () => void, inputs?: readonly unknown[]) => void;
+  useMemo: <TType>(factory: () => TType, inputs: readonly unknown[]) => TType;
+  useState: <TType>(
+    initialState: TType
+  ) => [TType, (newState: TType) => unknown];
+};
+
+type UseStorageParams<TItem extends Record<string, unknown>> = {
+  showRecent?: boolean | object;
+  query?: string;
+} & Pick<Parameters<UsePropGetters<TItem>>[0], 'indices' | 'indicesConfig'>;
+
+export function createAutocompleteStorage({
+  useEffect,
+  useMemo,
+  useState,
+}: CreateAutocompleteStorageParams) {
+  return function useStorage<TItem extends Record<string, unknown>>({
+    showRecent,
+    query,
+    indices,
+    indicesConfig,
+  }: UseStorageParams<TItem>) {
+    const storage = useMemo(() => createStorage({ limit: 5 }), []);
+    const [snapshot, setSnapshot] = useState(storage.getSnapshot());
+    useEffect(() => {
+      storage.registerUpdateListener(() => {
+        setSnapshot(storage.getSnapshot());
+      });
+      return () => {
+        storage.unregisterUpdateListener();
+      };
+    }, [storage]);
+
+    const storageHits = snapshot.getAll(query).map((value) => ({
+      objectID: value,
+      query: value,
+      __indexName: 'recent-searches',
+    }));
+
+    const indicesForPropGetters = [...indices];
+    const indicesConfigForPropGetters = [...indicesConfig];
+    if (showRecent) {
+      indicesForPropGetters.unshift({
+        indexName: 'recent-searches',
+        indexId: 'recent-searches',
+        hits: storageHits,
+      });
+      indicesConfigForPropGetters.unshift({
+        indexName: 'recent-searches',
+        // @ts-expect-error - we know it has query as it's generated from storageHits
+        getQuery: (item) => item.query,
+      });
+    }
+
+    return {
+      storage,
+      storageHits,
+      indicesForPropGetters,
+      indicesConfigForPropGetters,
+    };
+  };
+}
+
+const LOCAL_STORAGE_KEY_TEST = 'test-localstorage-support';
+const LOCAL_STORAGE_KEY = 'autocomplete-recent-searches';
```

Comment:
Is there some way where we link the autocomplete setup with the recent searches that are being presented? Do we do that in autocomplete.js? 

What if there's multiple autocomplete instances in multiple places in the app?

Reply:
I think we could but the original lib doesn't, even though it does have code to handle multiple autocomplete components on a page (which I don't think anyone needs really)